src/main/java/io/anserini/util/TrainingDataGenerator.java

```diff
@@ -0,0 +1,360 @@
+  }
+
+  private void initializeIndex() throws IOException {
+    if (kbIndexDirectory == null && args.index != null) {
+      LOG.info("Initializing KB index from directory: {}", args.index);
+      kbIndexDirectory = FSDirectory.open(Paths.get(args.index));
+    }
+  }
+
+  /**
+   * Gets index reader from the directory specified in the user parameters
+   *
+   * @return KB index reader
+   * @throws IOException on error
+   */
+  private IndexReader getKbIndexReader() throws IOException {
+    if (kbIndexReader == null) {
+      initializeIndex();
+      kbIndexReader = DirectoryReader.open(kbIndexDirectory);
+    }
+
+    return kbIndexReader;
+  }
+
+  /**
+   * @return index searcher
+   * @throws IOException on error
+   */
+  IndexSearcher getKbIndexSearcher() throws IOException {
+    if (kbIndexSearcher == null) {
+      kbIndexSearcher = new IndexSearcher(getKbIndexReader());
+    }
+
+    return kbIndexSearcher;
+  }
+
+  /**
+   * Get the current default index analyzer or create it.
+   * Use EnglishAnalyzer as the default.
+   *
+   * @return kb index analyzer
+   * @throws Exception
+   */
+  Analyzer getKbIndexAnalyzer() throws Exception {
+    if (kbIndexAnalyzer == null) {
+      kbIndexAnalyzer = new EnglishAnalyzer();
+    }
+
+    return kbIndexAnalyzer;
+  }
+
+  /**
+   * Helper function. Extracts value from literal that has a type
+   * (whether the type is a language for a string literal or a basic data type
+   * like date, int, etc.)
+   *
+   * @param literalString the string representation of the literal, including its type
+   * @return value of the literal
+   */
+  static String extractValueFromTypedLiteralString(String literalString) {
+    return NTriplesUtil.parseLiteral(literalString, valueFactory).stringValue();
+  }
+
+  /**
+   * Helper function. Converts freebase URI to freebase mention id
+   *
+   * @param freebaseUri freebase uri, similar to
+   * @return freebase mention id
+   */
+  static String freebaseUriToFreebaseId(String freebaseUri) {
+    return freebaseUri.substring(freebaseUri.lastIndexOf('/')).replace('.', '/');
+  }
+
+  /**
+   * Writes an array as a line separated by TRAINING_DATA_SEPARATOR.
+   *
+   * @param l    the logger to write to (the training data file)
+   * @param data array of data
+   */
+  void writeToTrainingFiles(Logger l, String... data) {
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < data.length; i++) {
+      if (i > 0)
+        sb.append(TRAINING_DATA_SEPARATOR);
+
+      if (data[i].contains(TRAINING_DATA_SEPARATOR))
+        sb.append("\"").append(data[i]).append("\"");
+      else
+        sb.append(data[i]);
+    }
+
+    l.info(sb.toString());
+  }
+
+  /**
+   * Determine which properties to retrieve and runs the training data retrieval pipeline.
+   *
+   * @throws Exception on error
+   */
+  private void generateTrainingData() throws Exception {
+    switch (args.property.toLowerCase()) {
+      case "birthdate":
+        birthdate();
```

Comment:
I prefer Java reflection to decide what method to be called at runtime.  Let's use [getMethod()](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getMethod-java.lang.String-java.lang.Class...-)  to decide what method to call based on the property selected by the user?